File: 2016/day19.py
# @formatter:off
from os.path import isfile, join as path_join
from sys import path as sys_path
sys_path.insert(1, path_join(sys_path[0], '..'))
file_name = path_join('input', 'day19.txt')
def to_list(mf: object = int, sep='\n'): return [mf(x) for x in open(file_name).read().split(sep) if x]
def to_gen(mf: object = int, sep='\n'): return (mf(x) for x in open(file_name).read().split(sep) if x)
def p1(*args): ans = part1(*args); print(ans); return ans
def p2(*args): ans = part2(*args); print(ans); return ans

if not isfile(file_name):
	from aoc import get_input_file
	get_input_file()
# @formatter:on

from utils import parsefile

elves = int(parsefile(file_name, None))


# elves = 5
# elves = 1001


class Elf:
	def __init__(self, id):
		self.id = id
		self.presents = 1
		self.prev = None
		self.next = None

	def take(self):
		self.presents += self.next.presents
		self.next = self.next.next


def part1():
	elf1 = Elf(1)

	prev_elf = elf1
	for i in range(1, elves):
		elf = Elf(i + 1)
		elf.prev = prev_elf
		prev_elf.next = elf
		prev_elf = elf

	# fix last and first
	prev_elf.next = elf1
	elf1.prev = prev_elf

	curr_elf = elf1
	while curr_elf.next != curr_elf:
		curr_elf.take()
		curr_elf = curr_elf.next

	return curr_elf.id


from sortedcontainers import SortedList


class Elf_v2:
	count = 0
	elf_arr = []
	skiplist = None
	sortedlist = None
	elfdict = None

	def __init__(self, id):
		self.id = id
		self.presents = 1
		self.prev = None
		self.next = None

	def take(self):
		steal_count = Elf_v2.count // 2
		curr_index = Elf_v2.sortedlist.index(self.id)

		steal_index = (curr_index + steal_count) % Elf_v2.count

		steal_id = Elf_v2.sortedlist.pop(steal_index)
		steal_from = Elf_v2.elfdict[steal_id]

		self.presents += steal_from.presents
		steal_from.prev.next = steal_from.next
		steal_from.next.prev = steal_from.prev

		Elf_v2.count -= 1


def part2():
	Elf_v2.count = elves

	sortedlist = SortedList()
	elfdict = {}

	elf1 = Elf_v2(1)
	sortedlist.add(1)
	elfdict[1] = elf1

	elf_arr = [None] * elves
	Elf_v2.elf_arr = elf_arr

	elf_arr[0] = elf1

	prev_elf = elf1
	for i in range(1, elves):
		elf = Elf_v2(i + 1)
		sortedlist.add(i + 1)
		elfdict[i + 1] = elf
		elf.prev = prev_elf
		prev_elf.next = elf
		elf_arr[i] = elf
		prev_elf = elf

	# fix last and first
	prev_elf.next = elf1
	elf1.prev = prev_elf

	Elf_v2.sortedlist = sortedlist
	Elf_v2.elfdict = elfdict

	curr_elf = elf1
	while Elf_v2.count > 1:
		curr_elf.take()
		curr_elf = curr_elf.next

	curr_elf = curr_elf.next

	return curr_elf.id


p1()
p2()

File: cpp/2016/day19.cpp
#include "../aocHelper.h"

#include "skiplist/SkipList.h"

class Elf
{
public:
	int id;
	int presents = 1;
	Elf* prev = nullptr;
	Elf* next = nullptr;
	Elf(int id) : id(id) {}

	void take_v1()
	{
		this->presents += this->next->presents;
		this->next = this->next->next;
	}

	void take_v2(OrderedStructs::SkipList::HeadNode<int>& skiplist, Elf** elves, int elf_count)
	{
		int steal_count = elf_count / 2;
		int curr_index = skiplist.index(this->id);

		int steal_index = (curr_index + steal_count) % elf_count;

		int steal_id = skiplist.at(steal_index);
		Elf* steal_from = elves[steal_id - 1];

		this->presents += steal_from->presents;
		steal_from->prev->next = steal_from->next;
		steal_from->next->prev = steal_from->prev;

		skiplist.remove(steal_id);
	}
};

class Day19 : public BaseDay
{
public:
	Day19() : BaseDay("19") {}

	result_type solve() override
	{
		long long part1 = 0, part2 = 0;

		// parse input
		int elf_count = numericParse<int>(input);

		// create the elves
		Elf** elves = new Elf * [elf_count];

		for (int i = 0; i < elf_count; i++)
		{
			elves[i] = new Elf(i + 1);
		}

		// part 1

		// setup pointers
		Elf* elf1 = elves[0];
		Elf* prev_elf = elf1;
		for (int i = 1; i < elf_count; i++)
		{
			Elf* curr_elf = elves[i];
			curr_elf->prev = prev_elf;
			prev_elf->next = curr_elf;
			prev_elf = curr_elf;
		}
		prev_elf->next = elf1;
		elf1->prev = prev_elf;

		Elf* curr_elf = elf1;
		while (curr_elf != curr_elf->next)
		{
			curr_elf->take_v1();
			curr_elf = curr_elf->next;
		}

		part1 = curr_elf->id;


		// part 2
		int current_elf_count = elf_count;

		OrderedStructs::SkipList::HeadNode<int> skiplist;

		elf1 = elves[0];
		prev_elf = elf1;
		skiplist.insert(elf1->id);
		for (int i = 1; i < elf_count; i++)
		{
			Elf* curr_elf = elves[i];
			curr_elf->prev = prev_elf;
			prev_elf->next = curr_elf;
			skiplist.insert(curr_elf->id);
			prev_elf = curr_elf;
		}
		prev_elf->next = elf1;
		elf1->prev = prev_elf;


		curr_elf = elf1;
		while (curr_elf != curr_elf->next)
		{
			curr_elf->take_v2(skiplist, elves, current_elf_count);
			curr_elf = curr_elf->next;
			current_elf_count--;
		}

		part2 = curr_elf->id;

		// delete the elves
		for (int i = 0; i < elf_count; i++)
		{
			delete elves[i];
		}

		delete[] elves;

		return { part1, part2 };
	}
};
